com.unity.ml-agents/Runtime/InplaceArray.cs

```diff
@@ -0,0 +1,250 @@
+            }
+        }
+
+        /// <summary>
+        /// Per-element access.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <exception cref="IndexOutOfRangeException"></exception>
+        public T this[int index]
+        {
+            get
+            {
+                if (index >= Length)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                switch (index)
+                {
+                    case 0:
+                        return m_Elem0;
+                    case 1:
+                        return m_Elem1;
+                    case 2:
+                        return m_Elem2;
+                    case 3:
+                        return m_Elem3;
+                    default:
+                        throw new IndexOutOfRangeException();
+                }
+            }
+
+            set
+            {
+                if (index >= Length)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                switch (index)
+                {
+                    case 0:
+                        m_Elem0 = value;
+                        break;
+                    case 1:
+                        m_Elem1 = value;
+                        break;
+                    case 2:
+                        m_Elem2 = value;
+                        break;
+                    case 3:
+                        m_Elem3 = value;
+                        break;
+                    default:
+                        throw new IndexOutOfRangeException();
+                }
+            }
+        }
+
+        /// <summary>
+        /// The length of the array.
+        /// </summary>
+        public int Length
+        {
+            get => m_Length;
+        }
+
+        /// <summary>
+        /// Returns a string representation of the array's elements.
+        /// </summary>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException"></exception>
+        public override string ToString()
+        {
+            switch (m_Length)
+            {
+                case 1:
+                    return $"[{m_Elem0}]";
+                case 2:
+                    return $"[{m_Elem0}, {m_Elem1}]";
+                case 3:
+                    return $"[{m_Elem0}, {m_Elem1}, {m_Elem2}]";
+                case 4:
+                    return $"[{m_Elem0}, {m_Elem1}, {m_Elem2}, {m_Elem3}]";
+                default:
+                    throw new IndexOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Check that the arrays have the same length and have all equal values.
+        /// </summary>
+        /// <param name="lhs"></param>
+        /// <param name="rhs"></param>
+        /// <returns>Whether the arrays are equivalent.</returns>
+        public static bool operator==(InplaceArray<T> lhs, InplaceArray<T> rhs)
```

Comment:
What's the recommended way to do operator== and Equals() for a value type? Should the implementation be in Equals() instead?